cranelift/codegen/src/isa/s390x/abi.rs

```diff
@@ -607,28 +605,35 @@ impl ABIMachineSpec for S390xMachineDeps {
 
     fn gen_clobber_restore(
         call_conv: isa::CallConv,
-        _: &Signature,
-        _: &settings::Flags,
+        sig: &Signature,
+        flags: &settings::Flags,
         clobbers: &[Writable<RealReg>],
         fixed_frame_storage_size: u32,
         outgoing_args_size: u32,
     ) -> SmallVec<[Inst; 16]> {
         let mut insts = SmallVec::new();
+        let clobbered_callee_saves =
+            Self::get_clobbered_callee_saves(call_conv, flags, sig, clobbers);
 
         // Collect clobbered registers.
-        let (mut clobbered_gpr, clobbered_fpr) = get_regs_saved_in_prologue(call_conv, clobbers);
+        let mut clobbered_fpr = vec![];
+        let mut first_clobbered_gpr = 16;
         // We need to restore the link register in non-leaf functions.
         // FIXME: This should be included in the clobber list to begin with,
         // but isn't because we have have excluded call instructions via the
         // is_included_in_clobbers callback.
         if outgoing_args_size > 0 {
-            clobbered_gpr.push(Writable::from_reg(RealReg::from(gpr_preg(14))));
+            first_clobbered_gpr = 14;
         }
-        let mut first_clobbered_gpr = 16;
-        for reg in clobbered_gpr {
-            let enc = reg.to_reg().hw_enc();
-            if enc < first_clobbered_gpr {
-                first_clobbered_gpr = enc;
+        for &reg in clobbered_callee_saves.iter() {
```

Comment:
Could we factor out this loop and link-register conditional (from here and `gen_clobber_save` above) so we have `let (clobbered_fpr, first_clobbered_gpr) = Self::get_clobbered_fpr_gpr(&clobbered_callee_saves[..], outgoing_args_size)` ?